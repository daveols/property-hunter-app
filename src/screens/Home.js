import React, {useState, useEffect} from 'react';
import {CommonActions} from '@react-navigation/native';

import {StyleSheet, FlatList} from 'react-native';
import {Chip, Subheading, ActivityIndicator} from 'react-native-paper';

import {useListingsNearSuburbSearch} from '../hooks';

import ListingCard from '../components/ListingCard';

const Home = ({navigation, route: {params}}) => {
  const [prevSelectedSuburb, setPrevSelectedSuburb] = useState();
  const {
    search: searchListings,
    loading: isLoadingListings,
    listings,
  } = useListingsNearSuburbSearch();

  const selectedSuburb =
    params && params.selectedSuburb ? params.selectedSuburb : undefined;

  useEffect(() => {
    if (selectedSuburb !== prevSelectedSuburb) {
      setPrevSelectedSuburb(selectedSuburb);
      searchListings(selectedSuburb);
    }
  }, [prevSelectedSuburb, searchListings, selectedSuburb]);

  const clearSuburb = () => {
    navigation.dispatch(CommonActions.setParams({selectedSuburb: undefined}));
  };

  return (
    <>
      {selectedSuburb ? (
        <>
          <Chip onClose={clearSuburb} style={styles.suburb}>
            {selectedSuburb.name}
          </Chip>
          {isLoadingListings ? (
            <ActivityIndicator size="large" />
          ) : (
            <FlatList
              data={listings}
              renderItem={({item}) => <ListingCard {...item.listing} />}
              keyExtractor={(_, i) => `listing-${i}`}
            />
          )}
        </>
      ) : (
        <Subheading style={styles.noSuburbHeadline}>
          Search suburbs to find your dream home!
        </Subheading>
      )}
    </>
  );
};

const styles = StyleSheet.create({
  suburb: {
    alignSelf: 'flex-start',
    margin: 16,
  },
  noSuburbHeadline: {
    alignSelf: 'center',
    textAlign: 'center',
    margin: 16,
    marginTop: 64,
  },
});

export default Home;
