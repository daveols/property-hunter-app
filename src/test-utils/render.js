import React from 'react';
import {render} from '@testing-library/react-native';

import {DomainAuthProvider} from 'contexts/DomainAuthContext';

export default (ui, {...options} = {}) => {
  const Providers = ({children}) => (
    <DomainAuthProvider>{children}</DomainAuthProvider>
  );

  return {...render(ui, {wrapper: Providers, ...options})};
};
