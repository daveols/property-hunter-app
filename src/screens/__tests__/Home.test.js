import React from 'react';

import render from 'test-utils/render';
import {fireEvent, wait} from '@testing-library/react-native';

import {CommonActions} from '@react-navigation/native';

import Home from 'screens/Home';

const mockListings = [
  {
    listing: {
      propertyDetails: {
        streetNumber: '69',
        street: 'Hosier Lane',
        suburb: 'Melbourne',
        bedrooms: 6,
        bathrooms: 9,
      },
      media: [
        {
          category: 'Image',
          url: 'https://sweetmelbournepics.com.au/hosier-lane.png',
        },
      ],
    },
  },
];

const baseProps = {
  navigation: {
    dispatch: jest.fn(),
  },
  route: {
    params: undefined,
  },
};

const propsWithSuburb = {
  ...baseProps,
  route: {
    params: {
      selectedSuburb: {name: 'Melbourne', state: 'vic', postcode: '3000'},
    },
  },
};

describe('Home', () => {
  test('renders correctly with no listings', () => {
    const {container, queryByText} = render(<Home {...baseProps} />);

    expect(queryByText(/find your dream home/i)).toBeTruthy();

    expect(container).toMatchSnapshot();
  });

  test('displays listings from api when suburb is added to params, then clears listings with button', async () => {
    fetch
      .once(JSON.stringify({access_token: '696969', expires_in: Infinity}))
      .once(JSON.stringify(mockListings));

    const {container, rerender, queryByText, getByText} = render(
      <Home {...baseProps} />,
    );
    rerender(<Home {...propsWithSuburb} />);

    await wait(() => expect(queryByText(/hosier lane/i)).toBeTruthy());

    expect(fetch.mock.calls).toMatchSnapshot();
    expect(container).toMatchSnapshot();

    // dirty hack to get the close icon touchable
    const clearButton = getByText('ï…™');
    fireEvent.press(clearButton);

    expect(propsWithSuburb.navigation.dispatch).toBeCalledWith(
      CommonActions.setParams({selectedSuburb: undefined}),
    );
  });
});
