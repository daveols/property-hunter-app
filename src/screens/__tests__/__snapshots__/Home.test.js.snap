// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home displays listings from api when suburb is added to params, then clears listings with button 1`] = `
Array [
  Array [
    "https://auth.domain.com.au/v1/connect/token",
    Object {
      "body": "grant_type=client_credentials&scope=api_listings_read",
      "headers": Object {
        "authorization": "Basic Y2xpZW50X2EzNjhhODllOGVkZjA3MjUxOGQwMmM4MjcxNjJmNWU5OnNlY3JldF80YTRmMDVmMWEzZmIxNGI3MjY4MTQ3YTJhZGQ2NmNlMg==",
        "content-type": "application/x-www-form-urlencoded",
      },
      "method": "POST",
    },
  ],
  Array [
    "https://api.domain.com.au/v1/listings/residential/_search",
    Object {
      "body": "{\\"listingType\\":\\"Sale\\",\\"locations\\":[{\\"state\\":\\"VIC\\",\\"suburb\\":\\"Melbourne\\",\\"postCode\\":\\"3000\\",\\"includeSurroundingSuburbs\\":true}]}",
      "headers": Object {
        "Content-Type": "application/json",
        "authorization": "Bearer 696969",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`Home displays listings from api when suburb is added to params, then clears listings with button 2`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "flex-start",
        "backgroundColor": "#ebebeb",
        "borderColor": "#ebebeb",
        "borderRadius": 16,
        "borderStyle": "solid",
        "borderWidth": 0.5,
        "elevation": 0,
        "margin": 16,
      }
    }
  >
    <TouchableHighlight
      accessibilityComponentType="button"
      accessibilityRole="button"
      accessibilityStates={Array []}
      accessibilityTraits={
        Array [
          "button",
        ]
      }
      activeOpacity={0.85}
      delayPressIn={0}
      delayPressOut={100}
      disabled={true}
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 16,
          },
        ]
      }
      underlayColor="rgb(188, 188, 188)"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "paddingHorizontal": 4,
          }
        }
      >
        <Text
          numberOfLines={1}
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
                "textAlign": "left",
              },
              Array [
                Object {
                  "lineHeight": 24,
                  "marginVertical": 4,
                  "minHeight": 24,
                  "textAlignVertical": "center",
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "400",
                  "marginLeft": 8,
                  "marginRight": 4,
                },
                undefined,
              ],
            ]
          }
        >
          Melbourne
        </Text>
        <TouchableWithoutFeedback
          accessibilityComponentType="button"
          accessibilityTraits="button"
        >
          <View
            style={
              Object {
                "padding": 4,
              }
            }
          >
            <Text
              accessibilityElementsHidden={true}
              allowFontScaling={false}
              importantForAccessibility="no-hide-descendants"
              pointerEvents="none"
              style={
                Array [
                  Object {
                    "color": "rgba(0, 0, 0, 0.54)",
                    "fontSize": 16,
                  },
                  Array [
                    Object {
                      "transform": Array [
                        Object {
                          "scaleX": 1,
                        },
                      ],
                    },
                    Object {
                      "backgroundColor": "transparent",
                    },
                  ],
                  Object {
                    "fontFamily": "Material Design Icons",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
        </TouchableWithoutFeedback>
      </View>
    </TouchableHighlight>
  </View>
  <ScrollView
    data={
      Array [
        Object {
          "listing": Object {
            "media": Array [
              Object {
                "category": "Image",
                "url": "https://sweetmelbournepics.com.au/hosier-lane.png",
              },
            ],
            "propertyDetails": Object {
              "bathrooms": 9,
              "bedrooms": 6,
              "street": "Hosier Lane",
              "streetNumber": "69",
              "suburb": "Melbourne",
            },
          },
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReachedThreshold={2}
    removeClippedSubviews={false}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View
      style={null}
    >
      <View
        style={
          Object {
            "backgroundColor": "#ffffff",
            "borderRadius": 4,
            "elevation": 1,
            "margin": 16,
            "marginVertical": 8,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.5,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 0.75,
          }
        }
      >
        <TouchableWithoutFeedback
          delayPressIn={0}
          disabled={true}
        >
          <View
            style={
              Object {
                "flexGrow": 1,
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "backgroundColor": "#eeeeee",
                    "height": 195,
                    "overflow": "hidden",
                  },
                  Object {
                    "borderTopLeftRadius": 4,
                    "borderTopRightRadius": 4,
                  },
                  undefined,
                ]
              }
            >
              <Image
                siblings={
                  Array [
                    "withTheme(Card.Cover)",
                    "withTheme(Card.Title)",
                    "Card.Content",
                  ]
                }
                source={
                  Object {
                    "uri": "https://sweetmelbournepics.com.au/hosier-lane.png",
                  }
                }
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "height": undefined,
                      "justifyContent": "flex-end",
                      "padding": 16,
                      "resizeMode": "cover",
                      "width": undefined,
                    },
                    Object {
                      "borderTopLeftRadius": 4,
                      "borderTopRightRadius": 4,
                    },
                  ]
                }
              />
            </View>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "paddingLeft": 16,
                  },
                  Object {
                    "height": 72,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "flex": 1,
                      "flexDirection": "column",
                      "height": 50,
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "color": "rgba(0, 0, 0, 0.87)",
                          "fontFamily": "System",
                          "fontWeight": "500",
                          "textAlign": "left",
                          "writingDirection": "ltr",
                        },
                        Array [
                          Object {
                            "fontSize": 20,
                            "letterSpacing": 0.15,
                            "lineHeight": 30,
                            "marginVertical": 2,
                          },
                          Array [
                            Object {
                              "minHeight": 30,
                            },
                            Object {
                              "marginBottom": 0,
                            },
                            undefined,
                          ],
                        ],
                        Array [
                          Object {
                            "fontSize": 20,
                            "letterSpacing": 0.15,
                            "lineHeight": 30,
                            "marginVertical": 2,
                          },
                          Array [
                            Object {
                              "minHeight": 30,
                            },
                            Object {
                              "marginBottom": 0,
                            },
                            undefined,
                          ],
                        ],
                      ],
                    ]
                  }
                >
                  69 Hosier Lane
                </Text>
                <Text
                  numberOfLines={1}
                  style={
                    Array [
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "textAlign": "left",
                      },
                      Array [
                        Object {
                          "color": "rgba(0, 0, 0, 0.54)",
                          "fontFamily": "System",
                          "fontWeight": "400",
                          "textAlign": "left",
                          "writingDirection": "ltr",
                        },
                        Array [
                          Object {
                            "fontSize": 12,
                            "letterSpacing": 0.4,
                            "lineHeight": 20,
                            "marginVertical": 2,
                          },
                          Array [
                            Object {
                              "marginVertical": 0,
                              "minHeight": 20,
                            },
                            undefined,
                          ],
                        ],
                        Array [
                          Object {
                            "fontSize": 12,
                            "letterSpacing": 0.4,
                            "lineHeight": 20,
                            "marginVertical": 2,
                          },
                          Array [
                            Object {
                              "marginVertical": 0,
                              "minHeight": 20,
                            },
                            undefined,
                          ],
                        ],
                      ],
                    ]
                  }
                >
                  Melbourne
                </Text>
              </View>
              <View />
            </View>
            <View
              style={
                Array [
                  Object {
                    "paddingHorizontal": 16,
                  },
                  Object {
                    "paddingBottom": 16,
                  },
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#9e9e9e",
                      "fontSize": 16,
                    },
                    Object {
                      "marginRight": 4,
                    },
                    Object {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.54)",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "textAlign": "left",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontSize": 12,
                          "letterSpacing": 0.4,
                          "lineHeight": 20,
                          "marginVertical": 2,
                        },
                        Object {
                          "marginRight": 16,
                        },
                      ],
                      Array [
                        Object {
                          "fontSize": 12,
                          "letterSpacing": 0.4,
                          "lineHeight": 20,
                          "marginVertical": 2,
                        },
                        Object {
                          "marginRight": 16,
                        },
                      ],
                    ],
                  ]
                }
              >
                6
              </Text>
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#9e9e9e",
                      "fontSize": 16,
                    },
                    Object {
                      "marginRight": 4,
                    },
                    Object {
                      "fontFamily": "Material Design Icons",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "color": "rgba(0, 0, 0, 0.54)",
                        "fontFamily": "System",
                        "fontWeight": "400",
                        "textAlign": "left",
                        "writingDirection": "ltr",
                      },
                      Array [
                        Object {
                          "fontSize": 12,
                          "letterSpacing": 0.4,
                          "lineHeight": 20,
                          "marginVertical": 2,
                        },
                        Object {
                          "marginRight": 16,
                        },
                      ],
                      Array [
                        Object {
                          "fontSize": 12,
                          "letterSpacing": 0.4,
                          "lineHeight": 20,
                          "marginVertical": 2,
                        },
                        Object {
                          "marginRight": 16,
                        },
                      ],
                    ],
                  ]
                }
              >
                9
              </Text>
            </View>
          </View>
        </TouchableWithoutFeedback>
      </View>
    </View>
  </ScrollView>
</View>
`;

exports[`Home renders correctly with no listings 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
          "textAlign": "left",
        },
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.87)",
            "fontFamily": "System",
            "fontWeight": "400",
            "textAlign": "left",
            "writingDirection": "ltr",
          },
          Array [
            Object {
              "fontSize": 16,
              "letterSpacing": 0.5,
              "lineHeight": 24,
              "marginVertical": 2,
            },
            Object {
              "alignSelf": "center",
              "margin": 16,
              "marginTop": 64,
              "textAlign": "center",
            },
          ],
          Array [
            Object {
              "fontSize": 16,
              "letterSpacing": 0.5,
              "lineHeight": 24,
              "marginVertical": 2,
            },
            Object {
              "alignSelf": "center",
              "margin": 16,
              "marginTop": 64,
              "textAlign": "center",
            },
          ],
        ],
      ]
    }
  >
    Search suburbs to find your dream home!
  </Text>
</View>
`;
